keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,185 @@
+- Clients can't know how to deal with that, especially if they can't parse the
+  OpenAPI
+- Server can't understand the user intent and normalize the object properly
+
+## Motivation
+
+Currently, changing a value in an one-of type is difficult because the semantic
+is implicit, which means that nothing can be built to automatically fix unions,
+leading to many bugs and issues:
+- https://github.com/kubernetes/kubernetes/issues/35345
+- https://github.com/kubernetes/kubernetes/issues/24238
+- https://github.com/kubernetes/kubernetes/issues/34292
+- https://github.com/kubernetes/kubernetes/issues/6979
+- https://github.com/kubernetes/kubernetes/issues/33766
+- https://github.com/kubernetes/kubernetes/issues/24198
+- https://github.com/kubernetes/kubernetes/issues/60340
+
+And then, for other people:
+- https://github.com/rancher/rancher/issues/13584
+- https://github.com/helm/charts/pull/12319
+- https://github.com/EnMasseProject/enmasse/pull/1974
+- https://github.com/helm/charts/pull/11546
+- https://github.com/kubernetes/kubernetes/pull/35343
+
+Server-side [apply](http://features.k8s.io/555) is what enables this proposal to
+become possible.
+
+### Goals
+
+The goal is to enable a union or "oneof" semantic in Kubernetes types, both for
+in-tree types and for CRDs.
+
+### Non-Goals
+
+We're not planning to use this KEP to release the feature, but mostly as a way
+to document what we're doing.
+
+## Proposal
+
+In order to support unions in a backward compatible way in kubernetes, we're
+proposing the following changes.
+
+Note that this proposes unions to be "at most one of". Whether exactly one is
+supported or not should be implemented by the validation logic.
+
+### Go tags
+
+We're proposing a new type of tags for go types (in-tree types, and also
+kubebuilder types):
+
+- `// +oneof` before a field means that it is part of the union. Multiple fields
+  can have this prefix. These fields MUST BE optional, omitempty and pointers.
+- `// +discriminator` before a field means that this field is the discriminator
+  for the union. Only one field per structure can have this prefix. This field HAS
+  TO be a string, and CAN be optional.
+
+Note that this means that only one "union" type can be defined per structure,
+which is a deliberate decision. This also includes go embedding which, once
+flattened, would define multiple unions in the same structure. Also note that we
+have no occurrences of these existing in core-types (AFAIK).
+
+### OpenAPI
+
+OpenAPI v3 already allows a "oneOf" form, which is accepted by CRD validation
+(and will continue to be accepted in the future), they will be interpreted
+properly if possible (see CRD GA effort).
+
+Since OpenAPI v2 has no support for oneOf, a couple of kubernetes specific
+extensions will be added to OpenAPI:
+- `x-kubernetes-oneof-discriminator: true` is set for the discriminator field,
+- `x-kubernetes-oneof: "<discriminatedName>` is set for discriminated
```

Comment:
for easier interpretation of our v3 output, shouldn't be add these extensions there as well?

At best we come up with some oneOf pattern in OpenAPI which is pretty hard to detect. There will be more things to check than just the existence of the oneOf operator. E.g. oneOf of to patterns is no union.

Reply:
Another point: I don't want to see that we turn a concrete oneOf pattern into some API convention that we cannot change ever for compatibilty reasons. It is much easier to leave the pattern open (of course it must comply with OpenAPI semantics, but nobody should depend on it to stay the same), and at the same time have these more abstract extensions which will stay forever and are part of the API server API contract.

Reply:
Last but not least, we have to offer these extensions in the CRD validation schemata. Translating some oneOf pattern into those extensions for v2 is unnecessary complex. Also we couple the openapi generator to this pattern (compare previous comment), although both apiserver and generator have different life cycles.